aiplatform/api/AIPlatform.Samples/FunctionCalling.cs

```diff
@@ -0,0 +1,168 @@
+/*
+ * Copyright 2024 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// [START generativeaionvertexai_gemini_function_calling]
+
+using Google.Cloud.AIPlatform.V1;
+using System;
+using System.Threading.Tasks;
+using Type = Google.Cloud.AIPlatform.V1.Type;
+using Value = Google.Protobuf.WellKnownTypes.Value;
+
+public class FunctionCalling
+{
+    public async Task<string> GenerateFunctionCall(
+        string projectId = "your-project-id",
+        string location = "us-central1",
+        string publisher = "google",
+        string model = "gemini-1.0-pro-001")
+    {
+        var predictionServiceClient = new PredictionServiceClientBuilder
+        {
+            Endpoint = $"{location}-aiplatform.googleapis.com"
+        }.Build();
+
+        // Define the user's prompt in a Content object that we can reuse in
+        // model calls
+        var userPromptContent = new Content
+        {
+            Role = "USER",
+            Parts =
+            {
+                new Part { Text = "What is the weather like in Boston?" }
+            }
+        };
+
+        // Specify a function declaration and parameters for an API request
+        var functionName = "get_current_weather";
+        var getCurrentWeatherFunc = new FunctionDeclaration
+        {
+            Name = functionName,
+            Description = "Get the current weather in a given location",
+            Parameters = new OpenApiSchema
+            {
+                Type = Type.Object,
+                Properties =
+                {
+                    {"location", new()
+                        {
+                            Type = Type.String,
+                            Description = "Get the current weather in a given location"
+                        }
+                    },
+                    {"unit", new()
+                        {
+                            Type = Type.String,
+                            Enum = { "celsius", "fahrenheit" },
+                            Description = "The unit of measurement for the temperature"
+                        }
+                    }
```

Comment:
I think I'd do this instead. Wait for @jskeet in case he has a better suggestions.

```suggestion
                    {
                        "location",
                        new()
                        {
                            Type = Type.String,
                            Description = "Get the current weather in a given location"
                        }
                    },
                    {
                        "unit",
                        new()
                        {
                            Type = Type.String,
                            Enum = { "celsius", "fahrenheit" },
                            Description = "The unit of measurement for the temperature"
                        }
                    }
```

Reply:
I'd probably use indexer notation actually:

```csharp
["location"] = new() { Type = Type.String, Description = "Get the current weather in a given location" },
["unit"] = new()
{ 
    Type = Type.String,
    Description = "The unit of measurement for the temperature",
    Enum = {"celsius", "fahrenheit"
}
```

Reply:
Thanks, made the changes